src/Microsoft.AspNetCore.Blazor.Server/Circuits/CircuitSynchronizationContext.cs

```diff
@@ -0,0 +1,261 @@
+                    _state.IsBusy = false;
+                    Monitor.Exit(_state.Monitor);
+                }
+
+                return;
+            }
+
+            // If we get here is means that a callback is being queued while something is currently executing
+            // in this context. Let's instead add it to the queue and yield.
+            //
+            // We use our own queue here to maintain the execution order of the callbacks scheduled here. Also
+            // we need a queue rather than just scheduling an item in the thread pool - those items would immediately
+            // block and hurt scalability.
+            //
+            // We need to capture the execution context so we can restore it later. This code is similar to
+            // the call path of ThreadPool.QueueUserWorkItem and System.Threading.QueueUserWorkItemCallback.
+            ExecutionContext context = null;
+            if (!ExecutionContext.IsFlowSuppressed())
+            {
+                context = ExecutionContext.Capture();
+            }
+
+            NotifyPendingWork(new WorkItem()
+            {
+                Context = context,
+                Callback = d,
+                State = state,
+            });
+        }
+
+        // synchronously runs the callback
+        public override void Send(SendOrPostCallback d, object state)
+        {
+            try
+            {
+                Monitor.Enter(_state.Monitor);
+                _state.IsBusy = true;
+
+                ExecuteSynchronously(d, state);
+            }
+            finally
+            {
+                _state.IsBusy = false;
+                Monitor.Exit(_state.Monitor);
+            }
+        }
+
+        // shallow copy
+        public override SynchronizationContext CreateCopy()
+        {
+            return new CircuitSynchronizationContext(_state);
+        }
+
+        private void NotifyPendingWork(WorkItem item)
+        {
+            lock (_state.WorkerLock)
+            {
+                _state.Queue.Enqueue(item);
+
+                if (!_state.IsWorkerActive)
+                {
+                    _state.IsWorkerActive = true;
+
+                    // Perf - using a static thunk here to avoid a delegate allocation.
+                    ThreadPool.UnsafeQueueUserWorkItem(BackgroundWorkThunk, this);
+                }
+            }
+        }
+
+        private void ExecuteSynchronously(SendOrPostCallback d, object state)
+        {
+            d(state);
+        }
+
+        private void ExecuteSynchronously(WorkItem item)
+        {
+            if (item.Context == null)
+            {
+                item.Callback(item.State);
+                return;
+            }
+
+            // Perf - using a static thunk here to avoid a delegate allocation.
+            ExecutionContext.Run(item.Context, ExecutionContextThunk, item);
+        }
+
+        private void ExecuteBackgroundWork()
+        {
+            Monitor.Enter(_state.Monitor);
```

Comment:
Note that this might block. I'm debating changing this to tasks instead of QUWI to avoid this, but it will be much much more complicated. 

There's a pattern used increasingly often to keep an instance field holding a task that represents the 'current' execution, and simply call `ContinueWith` and replace the task with the new task. This enforces ordering pretty nicely and doesn't require blocking - I'm sure this would invoke more allocations in some common cases though so there's a tradeoff. 

Overall I'm more worried about the blocking. Threads are a limited resource, allocations much more so.

Reply:
Note that the System.Web sync-context uses task-chaining.

Reply:
I actually came here to say don’t block and mimic the system.web context of anything.